README.rst
==================================================================
covid-stats: Charts and visualisations for the COVID-19 spread in Bulgaria
==================================================================

TODO: Modify the whole file as necessary.

This is a "long description" file for the package that you are creating.
If you submit your package to PyPi, this text will be presented on the `public page <http://pypi.python.org/pypi/python_package_boilerplate>`_ of your package.

Note: This README has to be written using `reStructured Text <http://docutils.sourceforge.net/rst.html>`_, otherwise PyPi won't format it properly.

Installation
------------

The easiest way to install most Python packages is via ``easy_install`` or ``pip``::

    $ easy_install covid_stats

Usage
-----

TODO: This is a good place to start with a couple of concrete examples of how the package should be used.

The boilerplate code provides a dummy ``main`` function that prints out the word 'Hello'::

    >> from covidstats import main
    >> main()
    
When the package is installed via ``easy_install`` or ``pip`` this function will be bound to the ``covid_stats`` executable in the Python installation's ``bin`` directory (on Windows - the ``Scripts`` directory).


CHANGELOG.md
Version 1.14
------------
    - Update the region map coloring threshold values.

Version 1.13
------------
    - Update the discount of the pydlm model to obtain better results.

Version 1.12
------------
    - Added unvaccinated fatal percentage by age group plot.

Version 1.11
------------
    - Added infected by age group plot.
    - Added fatal by age group plot.
    - Added fatal percentage by age group plot.

Version 1.10
------------
    - Added monthly vaccinated fatal percentage plot.

Version 1.9
-----------
    - Added vaccinated infected by age group plot.
    - Added vaccinated hospitalized by age group plot.
    - Added vaccinated intensive care by age group plot.
    - Added vaccinated fatal by age group plot.
    - Added vaccinated fatal percentage by age group plot.
    - Added program startup argument for external charts.

Version 1.8
-----------
    - Add historical hospitalized and intensive care plot.

Version 1.7
-----------
    - Fix figure size being altered after some plots' generation.

Version 1.6
-----------
    - Fix vaccinations plots translations.

Version 1.5
-----------
    - Added daily and weekly vaccination timeline plots.

Version 1.4
-----------
    - Added date cases age and week cases age plots.

Version 1.3
-----------
    - Added rolling biweekly places cases plot.

Version 1.2
-----------
    - Fix antigen tests positivity plot data starting point.

Version 1.1
-----------
    - Introduced weekly positive tests percentage charts.

Version 1.0
-----------
    - Introduced i18n - added bg (Bulgarian) and en (English) locales.
    - Generate all plots with the newly introduced locales.

Version 0.7
-----------
    - Applied tight layout to keep plot proportions the same.

Version 0.6
-----------
    - Fixed daily positivity plot ticks.

Version 0.5
-----------
    - Fixed inconsistent styling across different plots.

Version 0.4
-----------
    - Added historical cases plot.
    - Captioned all plots with generation date.

Version 0.3
-----------
    - Added daily positivity plot.

Version 0.2
-----------
    - Fixed forecast line starting point in the 14 days forecast plot.
    - Reformatted code style.

Version 0.1
-----------
    - Initial version.

tests/__init__.py
'''
covid-stats: tests module.

Meant for use with py.test.
Organize tests into files, each named xxx_test.py
Read more here: http://pytest.org/

Copyright 2021, Veselin Stoyanov
Licensed under Attribution-NonCommercial-ShareAlike 4.0 International
'''

tests/basic_test.py
'''
covid-stats: Test module.

Meant for use with py.test.
Write each test as a function named test_<something>.
Read more here: http://pytest.org/

Copyright 2021, Veselin Stoyanov
Licensed under Attribution-NonCommercial-ShareAlike 4.0 International
'''

def test_example():
    assert True


covidstats/data.py
import pandas as pd
import datetime as dt

from covidstats import helpers


def get_week_cases_df():
    week_cases_df = pd.read_csv(
        'https://raw.githubusercontent.com/COVID-19-Bulgaria/covid-database/master/Bulgaria/WeekCasesDataset.csv')
    week_cases_df['date'] = week_cases_df.apply(
        lambda row: pd.to_datetime(dt.date.fromisocalendar(row.year, row.week, 7)), axis=1)

    return week_cases_df


def get_week_places_cases_df():
    week_places_cases_df = pd.read_csv(
        'https://raw.githubusercontent.com/COVID-19-Bulgaria/covid-database/master/Bulgaria/WeekPlacesCasesDataset.csv')
    week_places_cases_df['date'] = week_places_cases_df.apply(
        lambda row: dt.date.fromisocalendar(row.year, row.week, 7), axis=1)

    return week_places_cases_df


def get_active_cases_df():
    active_cases_dataset = pd.read_json(
        'https://raw.githubusercontent.com/COVID-19-Bulgaria/covid-database/master/Bulgaria/DateActiveCasesDataset.json'
    )
    active_cases_dataset['active'] = active_cases_dataset['active'].apply(pd.Series)

    return active_cases_dataset


def get_date_cases_df():
    date_cases_dataset = pd.read_json(
        'https://raw.githubusercontent.com/COVID-19-Bulgaria/covid-database/master/Bulgaria/DateCasesDataset.json')

    for column in ['infected', 'cured', 'fatal', 'hospitalized', 'intensive_care', 'medical_staff', 'pcr_tests',
                   'positive_pcr_tests', 'antigen_tests', 'positive_antigen_tests', 'vaccinated']:
        date_cases_dataset[column] = date_cases_dataset[column].apply(pd.Series)

    date_cases_dataset['date'] = date_cases_dataset.index

    return date_cases_dataset


def get_date_cases_age_df():
    date_cases_age_df = pd.read_csv('https://raw.githubusercontent.com/COVID-19-Bulgaria/covid-database/master/'
                                    'Bulgaria/CasesAgeDataset.csv', parse_dates=['date'])

    return date_cases_age_df


def build_date_diff_cases_age_df(df=get_date_cases_age_df()):
    date_diff_cases_age_df = pd.DataFrame({'date': df['date']})
    for column in ['group_0_19', 'group_20_29', 'group_30_39', 'group_40_49', 'group_50_59', 'group_60_69',
                   'group_70_79', 'group_80_89', 'group_90']:
        date_diff_cases_age_df[column] = df[column].diff()

    return date_diff_cases_age_df.dropna()


def get_date_diff_cases_df():
    date_diff_cases_dataset = pd.read_json(
        'https://raw.githubusercontent.com/COVID-19-Bulgaria/covid-database/master/Bulgaria/DateDiffCasesDataset.json')

    for column in ['infected', 'cured', 'fatal', 'hospitalized', 'intensive_care', 'medical_staff', 'pcr_tests',
                   'positive_pcr_tests', 'antigen_tests', 'positive_antigen_tests', 'vaccinated']:
        date_diff_cases_dataset[column] = date_diff_cases_dataset[column].apply(pd.Series)

    date_diff_cases_dataset['date'] = date_diff_cases_dataset.index

    return date_diff_cases_dataset


def get_date_positive_tests_df():
    date_positive_tests_df = pd.read_csv('https://raw.githubusercontent.com/COVID-19-Bulgaria/covid-database/master/'
                                         'Bulgaria/DatePositiveTestsDataset.csv', parse_dates=['date'])

    return date_positive_tests_df


def get_rolling_biweekly_places_cases_df():
    rolling_biweekly_places_cases_df = pd.read_csv('https://raw.githubusercontent.com/COVID-19-Bulgaria/covid-database/'
                                                   'master/Bulgaria/RollingBiWeeklyPlacesCasesDataset.csv',
                                                   parse_dates=['date'])

    return rolling_biweekly_places_cases_df


def get_data_egov_bg_df(resource):
    df = pd.read_csv(resource,
                     storage_options={
                         'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.'
                                       '36 (KHTML, like Gecko) Chrome/95.0.4638.69 Safari/537.36'},
                     parse_dates=['Дата'], cache_dates=True, encoding='utf-8')

    return df


def get_infected_by_age_group_df():
    infected_by_age_group_df = get_data_egov_bg_df('https://data.egov.bg/resource/download/'
                                                   '8f62cfcf-a979-46d4-8317-4e1ab9cbd6a8/csv')
    infected_by_age_group_df.rename(columns={'Дата': 'date'}, inplace=True)

    return infected_by_age_group_df


def get_fatal_by_age_group_df():
    fatal_by_age_group_df = get_data_egov_bg_df('https://data.egov.bg/resource/download/'
                                                '18851aca-4c9d-410d-8211-0b725a70bcfd/csv')
    rename_age_df_columns(fatal_by_age_group_df, 'Брой починали', 'fatal')

    return fatal_by_age_group_df


def rename_age_df_columns(df, column_bg, column_en):
    df.rename(columns={
        'Дата': 'date',
        'Пол': 'sex',
        'Възрастова група': 'age',
        column_bg: column_en
    }, inplace=True)


def build_total_infected_by_age_group_df(df):
    total_infected_by_age_group_df = df[df['date'] == df['date'].max()].drop(['date', '0 - 19'], axis=1)
    total_infected_by_age_group_df = total_infected_by_age_group_df.melt(var_name='age', value_name='infected')
    total_infected_by_age_group_df['infected'] = pd.to_numeric(total_infected_by_age_group_df['infected'])

    return total_infected_by_age_group_df


def aggregate_0_19_age_group(df):
    if len(df[df['age'] == '20 - 29'].index) == 0:
        df.loc[len(df.index)] = ['20 - 29', 0]
        df.sort_values(by='age', inplace=True, ignore_index=True)

    age_group_0_19 = df.iloc[lambda row: row.index < df[df['age'] == '20 - 29'].index[0]].sum()[1]
    rest_age_groups = df.iloc[lambda row: row.index >= df[df['age'] == '20 - 29'].index[0]].reset_index(drop=True)

    rest_age_groups.loc[len(rest_age_groups.index)] = ['0 - 19', age_group_0_19]
    rest_age_groups.sort_values(by='age', inplace=True, ignore_index=True)

    return rest_age_groups


def rename_vaccinated_df_columns(df, column_bg, column_en):
    df.rename(columns={
        'Дата': 'date',
        'Ваксина': 'vaccine',
        'Пол': 'sex',
        'Възрастова група': 'age',
        column_bg: column_en
    }, inplace=True)


def get_infected_vaccinated_df():
    infected_vaccinated_df = get_data_egov_bg_df('https://data.egov.bg/resource/download/'
                                                 'e9f795a8-0146-4cf0-9bd1-c0ba3d9aa124/csv')
    rename_vaccinated_df_columns(infected_vaccinated_df, 'Брой заразени', 'infected')

    return infected_vaccinated_df


def get_hospitalized_vaccinated_df():
    hospitalized_vaccinated_df = get_data_egov_bg_df('https://data.egov.bg/resource/download/'
                                                     '6fb4bfb1-f586-45af-8dd2-3385499c3664/csv')
    rename_vaccinated_df_columns(hospitalized_vaccinated_df, 'Брой хоспитализирани', 'hospitalized')

    return hospitalized_vaccinated_df


def get_intensive_care_vaccinated_df():
    intensive_care_vaccinated_df = get_data_egov_bg_df('https://data.egov.bg/resource/download/'
                                                       '218d49de-88a8-472a-9bb2-b2a373bd7ab4/csv')
    rename_vaccinated_df_columns(intensive_care_vaccinated_df, 'Брой в интензивно отделение', 'intensive_care')

    return intensive_care_vaccinated_df


def get_fatal_vaccinated_df():
    fatal_vaccinated_df = get_data_egov_bg_df('https://data.egov.bg/resource/download/'
                                              'e6a72183-28e0-486a-b4e4-b5db8b60a900/csv')
    rename_vaccinated_df_columns(fatal_vaccinated_df, 'Брой починали', 'fatal')

    return fatal_vaccinated_df


def build_grouped_by_age_df(df, filter_column):
    grouped_by_age_df = df[df[filter_column] != '-'].groupby('age').sum()
    grouped_by_age_df.reset_index(inplace=True)

    return grouped_by_age_df


def build_grouped_by_age_fatal_percentage_df(infected_by_age_df, fatal_by_age_df):
    grouped_by_age_fatal_percentage_df = pd.merge(infected_by_age_df, fatal_by_age_df, on='age', how='outer')
    grouped_by_age_fatal_percentage_df['fatal'] = grouped_by_age_fatal_percentage_df['fatal'].fillna(0)
    grouped_by_age_fatal_percentage_df['fatal_percentage'] = (
            grouped_by_age_fatal_percentage_df['fatal'] * 100 / grouped_by_age_fatal_percentage_df['infected'])\
        .round(decimals=2)

    return grouped_by_age_fatal_percentage_df


def build_date_vaccinated_fatal_df(fatal_vaccinated_df):
    date_vaccinated_fatal_df = fatal_vaccinated_df[fatal_vaccinated_df.vaccine != '-'].groupby(
        pd.Grouper(key='date', freq='D')).sum()
    date_vaccinated_fatal_df.reset_index(inplace=True)

    date_vaccinated_fatal_df.rename(columns={'fatal': 'fatal_vaccinated'}, inplace=True)

    return date_vaccinated_fatal_df


def build_vaccinated_fatal_percentage_df(date_vaccinated_fatal_df, date_diff_cases_df):
    vaccinated_fatal_percentage_df = pd.merge(date_vaccinated_fatal_df, date_diff_cases_df[['fatal', 'date']],
                                              on='date',
                                              how='outer')
    vaccinated_fatal_percentage_df = vaccinated_fatal_percentage_df[
        vaccinated_fatal_percentage_df['date'] >= pd.to_datetime('2021-01-01')]
    vaccinated_fatal_percentage_df['fatal_vaccinated'] = vaccinated_fatal_percentage_df['fatal_vaccinated'].fillna(0)
    vaccinated_fatal_percentage_df.sort_values(by='date', inplace=True)
    vaccinated_fatal_percentage_df = vaccinated_fatal_percentage_df.groupby(pd.Grouper(key='date', freq='M')).sum()
    vaccinated_fatal_percentage_df['fatal_vaccinated_percentage'] = vaccinated_fatal_percentage_df[
                                                                        'fatal_vaccinated'] * 100 / \
                                                                    vaccinated_fatal_percentage_df['fatal']

    return vaccinated_fatal_percentage_df


def build_rts_df(predicted_rts, start_date):
    df_index = pd.date_range(start_date, periods=len(predicted_rts), freq='D')

    increase_rts = [row + row * 0.05 for row in predicted_rts]
    decline_rts = [row - row * 0.05 for row in predicted_rts]

    df = pd.DataFrame({
        'predicted_rt': predicted_rts,
        'increase_rt': increase_rts,
        'decline_rt': decline_rts
    }, index=df_index)

    return df


def build_predicted_cases_df(reported_cases, rts_df, start_date):
    df_index = pd.date_range(start_date, periods=len(rts_df), freq='D')

    df = pd.DataFrame({
        'predicted_cases': helpers.predict_cases(reported_cases, rts_df['predicted_rt']),
        'increase_cases': helpers.predict_cases(reported_cases, rts_df['increase_rt']),
        'decline_cases': helpers.predict_cases(reported_cases, rts_df['decline_rt'])
    }, index=df_index)

    return df


def build_weekly_predicted_cases_df(week_cases_df, predicted_cases_df, start_date):
    df = predicted_cases_df.groupby(pd.Grouper(freq='W')).mean()
    df['predicted_cases'] = df['predicted_cases'].multiply(7)
    df['increase_cases'] = df['increase_cases'].multiply(7)
    df['decline_cases'] = df['decline_cases'].multiply(7)

    start_date_weekday = start_date.weekday()
    start_of_week_date = start_date - dt.timedelta(days=start_date_weekday)
    end_of_week_date = start_of_week_date + dt.timedelta(days=6)

    if end_of_week_date >= start_date:
        if week_cases_df.loc[week_cases_df['date'] == end_of_week_date].empty:
            current_cases = week_cases_df[week_cases_df['date'] <= end_of_week_date].iloc[-1]['infected']
            previous_week_date = end_of_week_date - dt.timedelta(weeks=1)
            df.loc[previous_week_date] = [current_cases, current_cases, current_cases]
            df.sort_index(inplace=True)
        else:
            current_cases = week_cases_df[week_cases_df['date'] == end_of_week_date].iloc[-1]['infected']
            average_week_cases = (current_cases / (start_date_weekday + 1)) * 7

            week_cases_df.loc[week_cases_df['date'] == end_of_week_date, ['infected']] = average_week_cases

            df.loc[df.index == end_of_week_date, ['predicted_cases']] = average_week_cases
            df.loc[df.index == end_of_week_date, ['increase_cases']] = average_week_cases
            df.loc[df.index == end_of_week_date, ['decline_cases']] = average_week_cases

    return df


def build_weekly_positive_tests_df(date_positive_tests_df=get_date_positive_tests_df()):
    weekly_positive_tests_df = date_positive_tests_df.groupby(pd.Grouper(key='date', freq='W')).agg({
        'pcr_tests': 'sum',
        'antigen_tests': 'sum',
        'positive_percentage': 'mean',
        'positive_pcr_tests': 'sum',
        'pcr_positive_percentage': 'mean',
        'positive_antigen_tests': 'sum',
        'antigen_positive_percentage': 'mean'
    })

    weekly_positive_tests_df['date'] = weekly_positive_tests_df.index
    weekly_positive_tests_df['total_tests'] = weekly_positive_tests_df['pcr_tests'] \
                                              + weekly_positive_tests_df['antigen_tests']
    weekly_positive_tests_df['total_positive_tests'] = weekly_positive_tests_df['positive_pcr_tests'] \
                                                       + weekly_positive_tests_df['positive_antigen_tests']

    return weekly_positive_tests_df


covidstats/__init__.py
'''
covid-stats: Main module

Copyright 2021, Veselin Stoyanov
Licensed under Attribution-NonCommercial-ShareAlike 4.0 International.
'''

from covidstats import data, plot, locales
import datetime as dt
import argparse

from covidstats.locales import t


def main():
    startup_arguments = get_startup_arguments()

    locales.setup_i18n()
    plot.setup_sns()

    if startup_arguments.external:
        build_external()
        return

    date_cases_df = data.get_date_cases_df()
    week_cases_df = data.get_week_cases_df()
    date_diff_cases_df = data.get_date_diff_cases_df()
    active_cases_df = data.get_active_cases_df()
    week_places_cases_df = data.get_week_places_cases_df()
    date_positive_tests_df = data.get_date_positive_tests_df()
    weekly_positive_tests_df = data.build_weekly_positive_tests_df(date_positive_tests_df)
    rolling_biweekly_places_cases_df = data.get_rolling_biweekly_places_cases_df()
    date_cases_age_df = data.get_date_cases_age_df()
    date_diff_cases_age_df = data.build_date_diff_cases_age_df(date_cases_age_df)

    generate_plots('bg', date_cases_df, week_cases_df, date_diff_cases_df, active_cases_df, week_places_cases_df,
                   date_positive_tests_df, weekly_positive_tests_df, rolling_biweekly_places_cases_df,
                   date_cases_age_df, date_diff_cases_age_df)

    generate_plots('en', date_cases_df, week_cases_df, date_diff_cases_df, active_cases_df, week_places_cases_df,
                   date_positive_tests_df, weekly_positive_tests_df, rolling_biweekly_places_cases_df,
                   date_cases_age_df, date_diff_cases_age_df)


def get_startup_arguments():
    parser = argparse.ArgumentParser()

    parser.add_argument('--external', action='store_true', default=False, dest='external',
                        help='Generate plots with external data.')

    return parser.parse_args()


def generate_plots(locale, date_cases_df, week_cases_df, date_diff_cases_df, active_cases_df, week_places_cases_df,
                   date_positive_cases_percentage_df, weekly_positive_tests_df, rolling_biweekly_places_cases_df,
                   date_cases_age_df, date_diff_cases_age_df):
    locales.set_locale(locale)

    # Weekly infected and cured cases plot
    infected_cured_week_cases_plot = plot.generate_week_cases_plot(
        df=week_cases_df,
        value_vars=['infected', 'cured'],
        hue_order=['infected', 'cured'],
        palette=['orange', 'green'],
        legend=[t('plots.week_cases_plot.legend.infected'), t('plots.week_cases_plot.legend.cured')]
    )
    plot.export_plot(infected_cured_week_cases_plot, '%s/WeeklyInfectedCured' % locale)

    # Weekly hospitalized, intensive care and fatal cases plot
    hospitalized_intensive_fatal_week_cases_plot = plot.generate_week_cases_plot(
        df=week_cases_df,
        value_vars=['hospitalized', 'intensive_care', 'fatal'],
        hue_order=['hospitalized', 'intensive_care', 'fatal'],
        palette=['pink', 'purple', 'red'],
        legend=[
            t('plots.week_cases_plot.legend.hospitalized'),
            t('plots.week_cases_plot.legend.intensive_care'),
            t('plots.week_cases_plot.legend.fatal')
        ]
    )
    plot.export_plot(hospitalized_intensive_fatal_week_cases_plot, '%s/WeeklyHospitalizedIntensiveCareFatal' % locale)

    # Weekly places cases plot
    places_cases_week_plot = plot.generate_week_places_cases_plot(week_places_cases_df)
    plot.export_plot(places_cases_week_plot, '%s/WeeklyPlacesCases' % locale)

    # 14 days forecast
    start_date = dt.date.today() + dt.timedelta(days=1)
    weekly_14_days_forecast_plot = plot.generate_weekly_14_days_prediction_plot_for_date(start_date, week_cases_df,
                                                                                         date_diff_cases_df)
    plot.export_plot(weekly_14_days_forecast_plot, '%s/Weekly14DaysForecast' % locale)

    # Active cases plot
    active_cases_plot = plot.generate_active_cases_plot(active_cases_df)
    plot.export_plot(active_cases_plot, '%s/ActiveCases' % locale)

    # Historical cases plot
    historical_cases_plot = plot.generate_combined_date_cases_plot(date_cases_df)
    plot.export_plot(historical_cases_plot, '%s/HistoricalCases' % locale)

    # Historical hospitalized and intensive care
    historical_hospitalized_intensive_care_cases_plot = plot.generate_date_cases_plot(
        df=date_cases_df, value_vars=['hospitalized', 'intensive_care'], hue_order=['hospitalized', 'intensive_care'],
        palette=['pink', 'purple'],
        legend=[
            t('plots.date_cases_plot.legend.hospitalized'),
            t('plots.date_cases_plot.legend.intensive_care'),
        ]
    )
    plot.export_plot(historical_hospitalized_intensive_care_cases_plot,
                     '%s/HistoricalHospitalizedIntensiveCareCases' % locale)

    # Daily positivity plot
    date_tests_positivity_plot = plot.generate_date_positive_cases_percentage_plot(
        date_positive_cases_percentage_df)
    plot.export_plot(date_tests_positivity_plot, '%s/DateTestsPositivity' % locale)

    # Weekly positivity plot
    weekly_tests_positivity_plot = plot.generate_tests_positivity_plot(
        df=weekly_positive_tests_df,
        value_vars=['total_tests', 'total_positive_tests'],
        hue_order=['total_tests', 'total_positive_tests'],
        main_palette=['orange', 'red'],
        main_legend=[
            t('plots.tests_positivity_plot.legend.tests'),
            t('plots.tests_positivity_plot.legend.positive_tests')
        ],
        secondary_var='positive_percentage',
        secondary_legend=t('plots.tests_positivity_plot.legend.positive_tests_percentage'),
        title=t('plots.tests_positivity_plot.title.pcr_antigen')
    )
    plot.export_plot(weekly_tests_positivity_plot, '%s/WeeklyTestsPositivity' % locale)

    # Weekly PCR positivity plot
    weekly_pcr_tests_positivity_plot = plot.generate_tests_positivity_plot(
        df=weekly_positive_tests_df,
        value_vars=['pcr_tests', 'positive_pcr_tests'],
        hue_order=['pcr_tests', 'positive_pcr_tests'],
        main_palette=['orange', 'red'],
        main_legend=[
            t('plots.tests_positivity_plot.legend.pcr_tests'),
            t('plots.tests_positivity_plot.legend.positive_pcr_tests')
        ],
        secondary_var='pcr_positive_percentage',
        secondary_legend=t('plots.tests_positivity_plot.legend.positive_tests_percentage'),
        title=t('plots.tests_positivity_plot.title.pcr')
    )
    plot.export_plot(weekly_pcr_tests_positivity_plot, '%s/WeeklyPCRTestsPositivity' % locale)

    # Weekly antigen positivity plot
    weekly_antigen_tests_positivity_plot = plot.generate_tests_positivity_plot(
        df=weekly_positive_tests_df[weekly_positive_tests_df.antigen_positive_percentage.notnull()],
        value_vars=['antigen_tests', 'positive_antigen_tests'],
        hue_order=['antigen_tests', 'positive_antigen_tests'],
        main_palette=['orange', 'red'],
        main_legend=[
            t('plots.tests_positivity_plot.legend.antigen_tests'),
            t('plots.tests_positivity_plot.legend.positive_antigen_tests')
        ],
        secondary_var='antigen_positive_percentage',
        secondary_legend=t('plots.tests_positivity_plot.legend.positive_tests_percentage'),
        title=t('plots.tests_positivity_plot.title.antigen')
    )
    plot.export_plot(weekly_antigen_tests_positivity_plot, '%s/WeeklyAntigenTestsPositivity' % locale)

    # Rolling biweekly places cases plot
    rolling_biweekly_places_cases_plot = plot.generate_rolling_biweekly_places_cases_facet_plot(
        rolling_biweekly_places_cases_df
    )
    plot.export_plot(rolling_biweekly_places_cases_plot, ('%s/RollingBiWeeklyPlacesCases' % locale), False)

    # Date cases age plot
    date_cases_age_plot = plot.generate_cases_age_plot(date_cases_age_df)
    plot.export_plot(date_cases_age_plot, '%s/DateCasesAge' % locale)

    # Week cases age plot
    date_week_cases_age_plot = plot.generate_week_cases_age_plot(date_diff_cases_age_df)
    plot.export_plot(date_week_cases_age_plot, '%s/WeekCasesAge' % locale)

    # Vaccination timeline by day
    date_vaccination_timeline_plot = plot.generate_vaccination_timeline_plot(df=date_cases_df,
                                                                             diff_df=date_diff_cases_df,
                                                                             plot_type='daily')
    plot.export_plot(date_vaccination_timeline_plot, '%s/DateVaccinationTimeline' % locale)

    # Vaccination timeline by week
    weekly_vaccination_timeline_plot = plot.generate_vaccination_timeline_plot(df=date_cases_df,
                                                                               diff_df=week_cases_df,
                                                                               plot_type='weekly')
    plot.export_plot(weekly_vaccination_timeline_plot, '%s/WeeklyVaccinationTimeline' % locale)


def build_external():
    infected_by_age_group_df = data.get_infected_by_age_group_df()
    fatal_by_age_group_df = data.get_fatal_by_age_group_df()
    infected_vaccinated_df = data.get_infected_vaccinated_df()
    hospitalized_vaccinated_df = data.get_hospitalized_vaccinated_df()
    intensive_care_vaccinated_df = data.get_intensive_care_vaccinated_df()
    fatal_vaccinated_df = data.get_fatal_vaccinated_df()

    total_infected_by_age_group_df = data.build_total_infected_by_age_group_df(infected_by_age_group_df)
    total_fatal_by_age_group_df = data.build_grouped_by_age_df(fatal_by_age_group_df, filter_column='age')
    grouped_by_age_fatal_percentage_df = data.build_grouped_by_age_fatal_percentage_df(
        data.aggregate_0_19_age_group(total_infected_by_age_group_df),
        data.aggregate_0_19_age_group(total_fatal_by_age_group_df)
    )
    infected_vaccinated_by_age_df = data.build_grouped_by_age_df(infected_vaccinated_df, filter_column='vaccine')
    hospitalized_vaccinated_by_age_df = data.build_grouped_by_age_df(hospitalized_vaccinated_df,
                                                                     filter_column='vaccine')
    intensive_care_vaccinated_by_age_df = data.build_grouped_by_age_df(intensive_care_vaccinated_df,
                                                                       filter_column='vaccine')
    fatal_vaccinated_by_age_df = data.build_grouped_by_age_df(fatal_vaccinated_df, filter_column='vaccine')
    vaccinated_by_age_fatal_percentage_df = data.build_grouped_by_age_fatal_percentage_df(
        infected_vaccinated_by_age_df, fatal_vaccinated_by_age_df)

    date_diff_cases_df = data.get_date_diff_cases_df()
    date_vaccinated_fatal_df = data.build_date_vaccinated_fatal_df(fatal_vaccinated_df)
    vaccinated_fatal_percentage_df = data.build_vaccinated_fatal_percentage_df(date_vaccinated_fatal_df,
                                                                               date_diff_cases_df)

    infected_unvaccinated_by_age_df = data.aggregate_0_19_age_group(total_infected_by_age_group_df).set_index('age')\
        .sub(data.aggregate_0_19_age_group(infected_vaccinated_by_age_df).set_index('age'), fill_value=0)
    fatal_unvaccinated_by_age_df = data.aggregate_0_19_age_group(total_fatal_by_age_group_df).set_index('age').sub(
        data.aggregate_0_19_age_group(fatal_vaccinated_by_age_df).set_index('age'), fill_value=0).reset_index()
    unvaccinated_by_age_fatal_percentage_df = data.build_grouped_by_age_fatal_percentage_df(
        infected_unvaccinated_by_age_df, fatal_unvaccinated_by_age_df)


    generate_external_plots('bg', total_infected_by_age_group_df, total_fatal_by_age_group_df,
                            grouped_by_age_fatal_percentage_df, infected_vaccinated_by_age_df,
                            hospitalized_vaccinated_by_age_df, intensive_care_vaccinated_by_age_df,
                            fatal_vaccinated_by_age_df, vaccinated_by_age_fatal_percentage_df,
                            vaccinated_fatal_percentage_df, unvaccinated_by_age_fatal_percentage_df)

    generate_external_plots('en', total_infected_by_age_group_df, total_fatal_by_age_group_df,
                            grouped_by_age_fatal_percentage_df, infected_vaccinated_by_age_df,
                            hospitalized_vaccinated_by_age_df, intensive_care_vaccinated_by_age_df,
                            fatal_vaccinated_by_age_df, vaccinated_by_age_fatal_percentage_df,
                            vaccinated_fatal_percentage_df, unvaccinated_by_age_fatal_percentage_df)


def generate_external_plots(locale, total_infected_by_age_group_df, total_fatal_by_age_group_df,
                            grouped_by_age_fatal_percentage_df, infected_vaccinated_by_age_df,
                            hospitalized_vaccinated_by_age_df, intensive_care_vaccinated_by_age_df,
                            fatal_vaccinated_by_age_df, vaccinated_by_age_fatal_percentage_df,
                            vaccinated_fatal_percentage_df, unvaccinated_by_age_fatal_percentage_df):
    locales.set_locale(locale)

    # Infected by age group
    total_infected_by_age_group_plot = plot.generate_grouped_by_age_bar_plot(
        total_infected_by_age_group_df,
        y='infected',
        color='orange',
        plot_type='total.infected'
    )
    plot.export_plot(total_infected_by_age_group_plot, '%s/InfectedByAgeGroup' % locale)

    # Fatal by age group
    total_fatal_by_age_group_plot = plot.generate_grouped_by_age_bar_plot(
        total_fatal_by_age_group_df,
        y='fatal',
        color='red',
        plot_type='total.fatal'
    )
    plot.export_plot(total_fatal_by_age_group_plot, '%s/FatalByAgeGroup' % locale)

    # Fatal percentage by age group
    grouped_by_age_group_fatal_percentage_plot = plot.generate_grouped_by_age_bar_plot(
        grouped_by_age_fatal_percentage_df,
        y='fatal_percentage',
        color='red',
        plot_type='total.fatal_percentage'
    )
    plot.export_plot(grouped_by_age_group_fatal_percentage_plot, '%s/FatalPercentageByAgeGroup' % locale)

    # Unvaccinated fatal percentage by age group
    unvaccinated_by_age_group_fatal_percentage_plot = plot.generate_grouped_by_age_bar_plot(
        unvaccinated_by_age_fatal_percentage_df,
        y='fatal_percentage',
        color='red',
        plot_type='unvaccinated.fatal_percentage'
    )
    plot.export_plot(unvaccinated_by_age_group_fatal_percentage_plot,
                     '%s/UnvaccinatedFatalPercentageByAgeGroup' % locale)

    # Vaccinated Infected By Age
    infected_vaccinated_by_age_plot = plot.generate_grouped_by_age_bar_plot(
        infected_vaccinated_by_age_df,
        y='infected',
        color='orange',
        plot_type='vaccinated.infected')
    plot.export_plot(infected_vaccinated_by_age_plot, '%s/VaccinatedByAgeInfected' % locale)

    # Vaccinated Hospitalized By Age
    hospitalized_vaccinated_by_age_plot = plot.generate_grouped_by_age_bar_plot(
        hospitalized_vaccinated_by_age_df,
        y='hospitalized',
        color='pink',
        plot_type='vaccinated.hospitalized')
    plot.export_plot(hospitalized_vaccinated_by_age_plot, '%s/VaccinatedByAgeHospitalized' % locale)

    # Vaccinated Intensive Care By Age
    intensive_care_vaccinated_by_age_plot = plot.generate_grouped_by_age_bar_plot(
        intensive_care_vaccinated_by_age_df,
        y='intensive_care',
        color='purple',
        plot_type='vaccinated.intensive_care')
    plot.export_plot(intensive_care_vaccinated_by_age_plot, '%s/VaccinatedByAgeIntensiveCare' % locale)

    # Vaccinated Fatal By Age
    fatal_vaccinated_by_age_plot = plot.generate_grouped_by_age_bar_plot(
        fatal_vaccinated_by_age_df,
        y='fatal',
        color='red',
        plot_type='vaccinated.fatal')
    plot.export_plot(fatal_vaccinated_by_age_plot, '%s/VaccinatedByAgeFatal' % locale)

    # Vaccinated Fatal Percentage By Age
    vaccinated_by_age_fatal_percentage_plot = plot.generate_grouped_by_age_bar_plot(
        vaccinated_by_age_fatal_percentage_df,
        y='fatal_percentage',
        color='red',
        plot_type='vaccinated.fatal_percentage')
    plot.export_plot(vaccinated_by_age_fatal_percentage_plot, '%s/VaccinatedByAgeFatalPercentage' % locale)

    # Monthly Vaccinated Fatal Percentage Plot
    vaccinated_fatal_percentage_plot = plot.generate_vaccinated_fatal_percentage_plot(vaccinated_fatal_percentage_df)
    plot.export_plot(vaccinated_fatal_percentage_plot, '%s/VaccinatedFatalPercentage' % locale)
